pkg/ring/kv/multi.go

```diff
@@ -0,0 +1,335 @@
+package kv
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/uber-go/atomic"
+	"golang.org/x/time/rate"
+
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+type multiConfig struct {
+	Primary   string
+	Secondary string
+
+	MirrorRateLimit float64
+	MirrorRateBurst int
+	MirrorEnabled   bool
+	MirrorPrefix    string
+
+	// Channel used by multiClient to get runtime config updates
+	RuntimeConfigChannel <-chan MultiRuntimeConfig
+
+	clients []kvclient
+}
+
+func (cfg *multiConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix string) {
+	f.StringVar(&cfg.Primary, prefix+"multi.primary", "", "Primary backend storage used by multi-client.")
+	f.StringVar(&cfg.Secondary, prefix+"multi.secondary", "", "Secondary backend storage used by multi-client.")
+	f.StringVar(&cfg.MirrorPrefix, prefix+"multi.mirror-prefix", "", "Prefix for keys that should be mirrored")
+	f.BoolVar(&cfg.MirrorEnabled, prefix+"multi.mirror-enabled", false, "Mirror values to secondary store")
+	f.Float64Var(&cfg.MirrorRateLimit, prefix+"multi.mirror-rate-limit", 1, "Rate limit used for mirroring. 0 disables rate limit.")
+	f.IntVar(&cfg.MirrorRateBurst, prefix+"multi.mirror-burst-size", 1, "Burst size used by rate limiter. Values less than 1 are treated as 1.")
+}
+
+// MultiRuntimeConfig has values that can change in runtime (via overrides)
+type MultiRuntimeConfig struct {
```

Comment:
I haven't noticed Validate before, thanks. I don't see much point in moving it there, as it would be the same code, just at a different place. If you insist, I don't mind, but I fail to see the value of it.

Reply:
Oh, this is not part of the standard config (it's in overrides). Please ignore my comment, and sorry for the noise.